File: python/42-Trapping-Rain-Water.py
## Note that there are two solutions
class Solution:
    def trap(self, height: List[int]) -> int:

        l , r = 0, len(height) -1
        maxl, maxr = height[l], height[r]
        res = 0

        while l < r:
            if maxl <= maxr:
                l+=1
                maxl = max(maxl, height[l])
                res+= maxl - height[l]
            else:
                r-=1
                maxr = max(maxr, height[r])
                res+= maxr - height[r]
        

        return res

## Second Solution

class Solution:
    def trap(self, height: List[int]) -> int:
        # formula is min[max height left, max hegiht right] - height[i]

        maxL = [0] * len(height)
        ml = 0
        maxR = [0] * len(height)
        mr = 0

        for i in range(len(height)):
            maxL[i] = ml 
            if height[i] > ml:
                ml = height[i]
        
        for i in range(len(height)-1 , -1 , -1):
            maxR[i] = mr 
            if height[i] > mr:
                mr = height[i]
        
        r = 0
        for i in range(len(height)):
            water = min(maxL[i], maxR[i]) - height[i]
            if water > 0:
                r+=water

        return r

File: java/42. Trapping Rain Water.java
//Neetcode - Two Pointer

class Solution {
    public int trap(int[] height) {
        /*
        To find out how much water can be trapped, it is important to know how water is trapped
        if index i has a building of height[i], then water can be trapped if:
            The lower of the building to its left or right is less than the height[i] 
                -> min(height[left], height[right]) < height[i]
            How much water trapped? its the difference of lowest left/right height and the height[i]
                -> min(height[left], height[right]) - height[i]
        Here, I implement a two pointer method to calculate the max height at left and right given a position i
        and then further calculate the actual water trapped.
        SOLUTION 2: https://leetcode.com/submissions/detail/792533678/
        */
        
        //declaring the two pointers
        int l = 0;
        int r = height.length -1;
        //setting max heights 
        int leftMax = height[l];
        int rightMax = height[r];
        //return result
        int res = 0;
        
        while (l<r){
            //here for the conditionals we want to move the lower height between left and right 
            //it is because the amount of water trapped is always limited by the lower not the higher height where it'd spill
            
            //moving the lower height and then we check the possibility of updating it 
            //then adding the difference (unit of water collected)
            if (leftMax < rightMax){
                l++;
                leftMax = Math.max(leftMax, height[l]);
                res += leftMax - height[l];
            } else{
                r--;
                rightMax = Math.max(rightMax, height[r]);
                res += rightMax - height[r];
            }
        }
        return res;
    }
}
